AppleSource/shell_cmds-198/id/id.1
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)id.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD: src/usr.bin/id/id.1,v 1.18 2006/09/26 22:28:12 ceri Exp $
.\"
.Dd September 26, 2006
.Dt ID 1
.Os
.Sh NAME
.Nm id
.Nd return user identity
.Sh SYNOPSIS
.Nm
.Op Ar user
.Nm
.Fl A
.Nm
.Fl F
.Op Ar user
.Nm
.Fl G Op Fl n
.Op Ar user
.Nm
.Fl M
.Nm
.Fl P
.Op Ar user
.Nm
.Fl g Op Fl nr
.Op Ar user
.Nm
.Fl p
.Op Ar user
.Nm
.Fl u Op Fl nr
.Op Ar user
.Sh DESCRIPTION
The
.Nm
utility displays the user and group names and numeric IDs, of the
calling process, to the standard output.
If the real and effective IDs are different, both are displayed,
otherwise only the real ID is displayed.
.Pp
If a
.Ar user
(login name or user ID)
is specified, the user and group IDs of that user are displayed.
In this case, the real and effective IDs are assumed to be the same.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl A
Display the process audit user ID and other process audit properties, which
requires privilege.
.It Fl F
Display the full name of the user.
.It Fl G
Display the different group IDs (effective, real and supplementary)
as white-space separated numbers, in no particular order.
.It Fl M
Display the MAC label of the current process.
.It Fl P
Display the id as a password file entry.
.It Fl a
Ignored for compatibility with other
.Nm
implementations.
.It Fl g
Display the effective group ID as a number.
.It Fl n
Display the name of the user or group ID for the
.Fl G ,
.Fl g
and
.Fl u
options instead of the number.
If any of the ID numbers cannot be mapped into names, the number will be
displayed as usual.
.It Fl p
Make the output human-readable.
If the user name returned by
.Xr getlogin 2
is different from the login name referenced by the user ID, the name
returned by
.Xr getlogin 2
is displayed, preceded by the keyword
.Dq login .
The user ID as a name is displayed, preceded by the keyword
.Dq uid .
If the effective user ID is different from the real user ID, the real user
ID is displayed as a name, preceded by the keyword
.Dq euid .
If the effective group ID is different from the real group ID, the real group
ID is displayed as a name, preceded by the keyword
.Dq rgid .
The list of groups to which the user belongs is then displayed as names,
preceded by the keyword
.Dq groups .
Each display is on a separate line.
.It Fl r
Display the real ID for the
.Fl g
and
.Fl u
options instead of the effective ID.
.It Fl u
Display the effective user ID as a number.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr who 1
.Sh STANDARDS
The
.Nm
function is expected to conform to
.St -p1003.2 .
.Sh HISTORY
The
historic
.Xr groups 1
command is equivalent to
.Dq Nm id Fl Gn Op Ar user .
.Pp
The
historic
.Xr whoami 1
command is equivalent to
.Dq Nm id Fl un .
.Pp
The
.Nm
command appeared in
.Bx 4.4 .


CHANGELOG.md
# Version 5.028.1
	- Fixed Cmd+v shortcut
	- HostKey fingerprint as base64 encoding
	- Smoothing the HUD on resize or SplitView
	- Empty default user configuration bugs.
	- Zoom shortcuts changed.
	- Solarized theme fixes.
	
# Version 4.024.2
	- AirTerminals! Put a terminal on your remote AirPlay screen :)
	- Blink shortcuts. Use your external keyboard to move, create, or remove terminals. Configure the trigger too! View settings > Keyboard > Shortcuts for more info.
	- Want to use SmartKeys when an external keyboard is connected? Now you can switch them on and off from configuration.
	- View geometry of the screen from the overlay with each resize.
	- Switch cursor blinking on/off from Appearance settings.

	- Fixed TouchID issues when returning to the app.
	- Fixed cursor blink sequence so that it always starts as ON.
	- Fixed Host Port not resetting properly.
	- Fixed overlay getting covered in landscape mode.

	- Smoother experience thanks to cleanups and improvements.

# Version 3.021.2
	- iCloud Hosts sync. Synchronize hosts between devices. If a Host already has been synced, it provides conflict resolution. No critical data like passwords is saved.
	- Auto Lock. If enabled, when you lock/unlock your device, Blink will also be locked. Passcode and TouchID will be required to unlock the app.
	- Added ARMv7 support. Support for 32 bit devices like iPad 2, 3, iPhone 5, etc... We will publish depending on how well it performs!
	- Added IPv6 support for hosts.
	- Share Public Encryption Keys. You can now share the public key from the Keys section to other apps, like Mail.

	- Updated Fira Code font to v1.204.
	- Improved error checking on Themes and Font uploads. Auto correct if the GH URL is not a raw one.

	- Fixed bug with password not getting saved on host creation.
	- Fixed crash when hitting arrows with landscape keyboard on Plus devices.
	- Rolled back LC_CTYPE enforcement on server.

# Version 2.109 / 3.0
	- NEW Autorepeat for normal keys on external keyboard. Vim users rejoice!
	- NEW Default User for connections based on device.
	- NEW Comments for Public Keys.

	- Security updates to libraries.

	- Fixed critical bugs affecting ssh connections.

# Version 1.031
	- NEW On-screen keyboard with more space for modifiers, FKeys and Cursor keys. Redesigned for more space on the modifiers, and with a central scrollable area that handles more keys. Activate the Alternate keys by taping on the Alt key. And now tap on a modifier to activate it as a normal button, or make a long press to chain different combinations.
	- NEW Add your own Fonts & Themes! More info on (https://github.com/blinksh/fonts) and (https://github.com/blinksh/themes)
	- NEW Multistep authentication. Servers with google authenticator or similar will now connect without problems :)
	- NEW Fira Code font with ligatures included.

	- Added -l parameter to ssh for Hosts.
	- Improved message on how use hosts after adding them.
	- Fixed on-screen arrows.
	- Fixed F0 as F10 on external keyboard.
	- Fixed keys selection problem on settings
	- Fixed ssh-copy-id issue when accessing the host for the first time

# Version 1.019
	- Simplified build process.
	- New README and BUILD instructions

# Version 0.1020
	- "Get In Touch" and "About" sections completed.
	- iOS 10 support.
	- Added exception for iOS distribution.

	- Improved gesture support, specially on iPhone and iOS10.
	- MBProgressHUD updated with patch.
	- Test Terminal fixes for fonts and updates.
	- Disable font ligatures.
	- SSH Session termination fixes.

# Version 0.1010
	- Terminal customization! Customize your terminal with different default themes and fonts. Preview your changes within the Settings Preview section. Save your changes and restore them on each execution. Enjoy!

	- Create an id_rsa on first boot as default key.

	- Swipe and pinch conflicts
	- Copy changes to settings

# Version 0.931
	- Map Caps or Shift taps to Esc key.
	- Cursor keys (Home, End, PgUp, PgDown) with default to Cmd+arrow.
	- Function keys (F1-F10) with default to Cmd+number.
	- Change mapping of any function key to a different combination.

	- Improved placeholders for Hosts configurations.

	- Removed Alt as Esc from initial defaults.

# Version 0.927
	This version should complete the experience in relation to ssh and terminal configurations, mimicking a big part of what you can do in a normal shell:
	- Hosts Configuration. Preconfigure a host parameters, like user, port, key and commands.
	- Connect to a Host by specifying its name. Do "mosh plankton"
	- Overwrite parameters from host configuration from the shell. So carlos@plankton will override the user field on plankton.
	- Default Modifier keys settings changed: Ctrl is Ctrl, and Alt sends Esc. Everything else is undefined.
	- Modifier keys configuration. CAPS as Ctrl or ESC, no problem! Cmd as Ctrl? You have it! Configure everything to you liking.
	- Added secure passwords stored on Keychain to Host Configuration.
	- Exiting the session within the MCP closes the Space.

	- Fixed hang after "exec request accepted". Establishing connections should be smooth now.
	- Fixed adjustments on viewport after rotating the display.
	- Fixed Ctrl + Space sequences.
	- Fixed wrong/unexisting Ctrl sequences.
	- Added Esc+any character support.

	- New settings added to project. Will start to fill up in raw branch.

# Version 0.916
	This version contains many important bug fixes that should improve the experience a lot and reduce crashes to, hopefully, close to zero:
	- Unicode support to Blink shell. Thanks Yury!
	- Support for function keys + modifiers. Now you can do Ctrl/Alt/Shift + Arrows.
	- Improved experience with drag down to close. Let me know what you think :)
	- SSH with improved function keys in ncurses apps. htop, less, etc... should now work perfectly.
	- New help message with version and gestures.
	- Added asterisk, underscore and escape to SmartKeys.
	- Added HockeySDKResources.bundle to project for improved updates.


AppleSource/shell_cmds-198/id/id.c
/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1991, 1993\n\r\
	The Regents of the University of California.  All rights reserved.\n\r";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)id.c	8.2 (Berkeley) 2/16/94";
#endif
#endif /* not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/usr.bin/id/id.c,v 1.33 2006/12/29 12:28:34 stefanf Exp $");

#include <sys/param.h>
#ifndef __APPLE__
#include <sys/mac.h>
#endif /* !__APPLE__ */

#ifdef USE_BSM_AUDIT
#include <bsm/audit.h>
#endif

#include <err.h>
#include <errno.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "error.h"

static void	id_print(struct passwd *, int, int, int);
static void	pline(struct passwd *);
static void	pretty(struct passwd *);
static void	auditid(void);
static void	fullname(struct passwd *);
static void	group(struct passwd *, int);
static void	maclabel(void);
static void	usage(void);
struct passwd *who(char *);

static int isgroups, iswhoami;

#ifdef __APPLE__
// SPI for 5235093
int32_t getgrouplist_2(const char *, gid_t, gid_t **);
#endif

int
id_main(int argc, char *argv[])
{
	struct group *gr;
	struct passwd *pw;
	int Gflag, Mflag, Pflag, ch, gflag, id, nflag, pflag, rflag, uflag;
	int Aflag;
	int Fflag;
	const char *myname;

	Gflag = Mflag = Pflag = gflag = nflag = pflag = rflag = uflag = 0;
	Aflag = 0;
	Fflag = 0;
    // Initialize:
    isgroups = 0;
    iswhoami = 0;

	myname = strrchr(argv[0], '/');
	myname = (myname != NULL) ? myname + 1 : argv[0];
	if (strcmp(myname, "groups") == 0) {
		isgroups = 1;
		Gflag = nflag = 1;
	}
	else if (strcmp(myname, "whoami") == 0) {
		iswhoami = 1;
		uflag = nflag = 1;
	}

	while ((ch = getopt(argc, argv,
	    (isgroups || iswhoami) ? "" : "AFPGMagnpru")) != -1)
		switch(ch) {
#ifdef USE_BSM_AUDIT
		case 'A':
			Aflag = 1;
			break;
#endif
		case 'F':
			Fflag = 1;
			break;
		case 'G':
			Gflag = 1;
			break;
		case 'M':
			Mflag = 1;
			break;
		case 'P':
			Pflag = 1;
			break;
		case 'a':
			break;
		case 'g':
			gflag = 1;
			break;
		case 'n':
			nflag = 1;
			break;
		case 'p':
			pflag = 1;
			break;
		case 'r':
			rflag = 1;
			break;
		case 'u':
			uflag = 1;
			break;
		case '?':
		default:
			usage();
            return 0;
		}
	argc -= optind;
	argv += optind;

    if (iswhoami && argc > 0) {
		usage();
        return 0;
    }

	switch(Aflag + Fflag + Gflag + Mflag + Pflag + gflag + pflag + uflag) {
	case 1:
		break;
	case 0:
		if (!nflag && !rflag)
			break;
		/* FALLTHROUGH */
	default:
		usage();
        return 0;
	}

	pw = *argv ? who(*argv) : NULL;

    if (Mflag && pw != NULL) {
		usage();
        return 0;
    }

#ifdef USE_BSM_AUDIT
	if (Aflag) {
		auditid();
		exit(0);
	}
#endif

	if (Fflag) {
		fullname(pw);
		exit(0);
	}

	if (gflag) {
		id = pw ? pw->pw_gid : rflag ? getgid() : getegid();
		if (nflag && (gr = getgrgid(id)))
			(void)printf("%s\n\r", gr->gr_name);
		else
			(void)printf("%u\n\r", id);
		exit(0);
	}

	if (uflag) {
		id = pw ? pw->pw_uid : rflag ? getuid() : geteuid();
		if (nflag && (pw = getpwuid(id)))
			(void)printf("%s\n\r", pw->pw_name);
		else
			(void)printf("%u\n\r", id);
		exit(0);
	}

	if (Gflag) {
		group(pw, nflag);
		exit(0);
	}

	if (Mflag) {
		maclabel();
		exit(0);
	}

	if (Pflag) {
		pline(pw);
		exit(0);
	}

	if (pflag) {
		pretty(pw);
		exit(0);
	}

	if (pw) {
		id_print(pw, 1, 0, 0);
	}
	else {
		id = getuid();
		pw = getpwuid(id);
		id_print(pw, 0, 1, 1);
	}
	exit(0);
}

void
pretty(struct passwd *pw)
{
	struct group *gr;
	u_int eid, rid;
	char *login;

	if (pw) {
		(void)printf("uid\t%s\n\r", pw->pw_name);
		(void)printf("groups\t");
		group(pw, 1);
	} else {
        if ((login = getlogin()) == NULL) {
			myerr(1, "getlogin");
            return;
        }

		pw = getpwuid(rid = getuid());
		if (pw == NULL || strcmp(login, pw->pw_name))
			(void)printf("login\t%s\n\r", login);
		if (pw)
			(void)printf("uid\t%s\n\r", pw->pw_name);
		else
			(void)printf("uid\t%u\n\r", rid);

		if ((eid = geteuid()) != rid) {
			if ((pw = getpwuid(eid)))
				(void)printf("euid\t%s\n\r", pw->pw_name);
			else
				(void)printf("euid\t%u\n\r", eid);
		}
		if ((rid = getgid()) != (eid = getegid())) {
			if ((gr = getgrgid(rid)))
				(void)printf("rgid\t%s\n\r", gr->gr_name);
			else
				(void)printf("rgid\t%u\n\r", rid);
		}
		(void)printf("groups\t");
		group(NULL, 1);
	}
}

void
id_print(struct passwd *pw, int use_ggl, int p_euid, int p_egid)
{
	struct group *gr;
	gid_t gid, egid, lastgid;
	uid_t uid, euid;
	int cnt, ngroups;
#ifdef __APPLE__
	gid_t *groups = NULL;
#else
	gid_t groups[NGROUPS + 1];
#endif
	const char *fmt;

#ifdef __APPLE__
	if (pw == NULL) {
		pw = getpwuid(getuid());
	}

	use_ggl = 1;
#endif

	if (pw != NULL) {
		uid = pw->pw_uid;
		gid = pw->pw_gid;
	}
	else {
		uid = getuid();
		gid = getgid();
	}

	if (use_ggl && pw != NULL) {
#ifdef __APPLE__
		// 5235093
		ngroups = getgrouplist_2(pw->pw_name, gid, &groups);
#else
		ngroups = NGROUPS + 1;
		getgrouplist(pw->pw_name, gid, groups, &ngroups);
#endif
	}
	else {
#ifdef __APPLE__
		groups = malloc((NGROUPS + 1) * sizeof(gid_t));
#endif
		ngroups = getgroups(NGROUPS + 1, groups);
	}

#ifdef __APPLE__
	if (ngroups < 0)
		warn("failed to retrieve group list");
#endif

	if (pw != NULL)
		printf("uid=%u(%s)", uid, pw->pw_name);
	else 
		printf("uid=%u", getuid());
	printf(" gid=%u", gid);
	if ((gr = getgrgid(gid)))
		(void)printf("(%s)", gr->gr_name);
	if (p_euid && (euid = geteuid()) != uid) {
		(void)printf(" euid=%u", euid);
		if ((pw = getpwuid(euid)))
			(void)printf("(%s)", pw->pw_name);
	}
	if (p_egid && (egid = getegid()) != gid) {
		(void)printf(" egid=%u", egid);
		if ((gr = getgrgid(egid)))
			(void)printf("(%s)", gr->gr_name);
	}
	fmt = " groups=%u";
	for (lastgid = -1, cnt = 0; cnt < ngroups; ++cnt) {
		if (lastgid == (gid = groups[cnt]))
			continue;
		printf(fmt, gid);
		fmt = ",%u";
		if ((gr = getgrgid(gid)))
			printf("(%s)", gr->gr_name);
		lastgid = gid;
	}
	printf("\n\r");
#ifdef __APPLE__
	free(groups);
#endif
}

#ifdef USE_BSM_AUDIT
void
auditid(void)
{
	auditinfo_addr_t auditinfo;

    if (getaudit_addr(&auditinfo, sizeof(auditinfo)) < 0) {
		myerr(1, "getaudit");
        return;
    }
	printf("auid=%d\n\r", auditinfo.ai_auid);
	printf("mask.success=0x%08x\n\r", auditinfo.ai_mask.am_success);
	printf("mask.failure=0x%08x\n\r", auditinfo.ai_mask.am_failure);
	printf("termid.port=0x%08x\n\r", auditinfo.ai_termid.at_port);
	printf("asid=%d\n\r", auditinfo.ai_asid);
}
#endif

void
fullname(struct passwd *pw)
{

	if (!pw) {
        if ((pw = getpwuid(getuid())) == NULL) {
			myerr(1, "getpwuid");
            return;
        }
	}

	(void)printf("%s\n\r", pw->pw_gecos);
}

void
group(struct passwd *pw, int nflag)
{
	struct group *gr;
	int cnt, id, lastid, ngroups;
#ifdef __APPLE__
	gid_t *groups = NULL;
#else
	gid_t groups[NGROUPS + 1];
#endif
	const char *fmt;

#ifdef __APPLE__
	if (pw == NULL) {
		pw = getpwuid(getuid());
	}
#endif

	if (pw) {
#ifdef __APPLE__
		// 5235093
		ngroups = getgrouplist_2(pw->pw_name, pw->pw_gid, &groups);
#else
		ngroups = NGROUPS + 1;
		(void) getgrouplist(pw->pw_name, pw->pw_gid, groups, &ngroups);
#endif
	} else {
#ifdef __APPLE__
		groups = malloc((NGROUPS + 1) * sizeof(gid_t));
#endif
		groups[0] = getgid();
		ngroups = getgroups(NGROUPS, groups + 1) + 1;
	}
	fmt = nflag ? "%s" : "%u";
	for (lastid = -1, cnt = 0; cnt < ngroups; ++cnt) {
		if (lastid == (id = groups[cnt]))
			continue;
		if (nflag) {
			if ((gr = getgrgid(id)))
				(void)printf(fmt, gr->gr_name);
			else
				(void)printf(*fmt == ' ' ? " %u" : "%u",
				    id);
			fmt = " %s";
		} else {
			(void)printf(fmt, id);
			fmt = " %u";
		}
		lastid = id;
	}
	(void)printf("\n\r");
#ifdef __APPLE__
	free(groups);
#endif
}

void
maclabel(void)
{
#ifdef __APPLE__
	myerrx(1, "-M unsupported");
    return;
#else /* !__APPLE__ */
	char *string;
	mac_t label;
	int error;

	error = mac_prepare_process_label(&label);
	if (error == -1)
		myerrx(1, "mac_prepare_type: %s", strerror(errno));

	error = mac_get_proc(label);
	if (error == -1)
		myerrx(1, "mac_get_proc: %s", strerror(errno));

	error = mac_to_text(label, &string);
	if (error == -1)
		myerrx(1, "mac_to_text: %s", strerror(errno));

	(void)printf("%s\n\r", string);
	mac_free(label);
	free(string);
#endif /* __APPLE__ */
}

struct passwd *
who(char *u)
{
	struct passwd *pw;
	long id;
	char *ep;

	/*
	 * Translate user argument into a pw pointer.  First, try to
	 * get it as specified.  If that fails, try it as a number.
	 */
	if ((pw = getpwnam(u)))
		return(pw);
	id = strtol(u, &ep, 10);
	if (*u && !*ep && (pw = getpwuid(id)))
		return(pw);
	myerrx(1, "%s: no such user", u);
    return NULL;
	/* NOTREACHED */
}

void
pline(struct passwd *pw)
{

	if (!pw) {
        if ((pw = getpwuid(getuid())) == NULL) {
			myerr(1, "getpwuid");
            return;
        }
	}

	(void)printf("%s:%s:%d:%d:%s:%ld:%ld:%s:%s:%s\n\r", pw->pw_name,
			pw->pw_passwd, pw->pw_uid, pw->pw_gid, pw->pw_class,
			(long)pw->pw_change, (long)pw->pw_expire, pw->pw_gecos,
			pw->pw_dir, pw->pw_shell);
}


void
usage(void)
{

	if (isgroups)
		(void)fprintf(stderr, "\rusage: groups [user]\n\r");
	else if (iswhoami)
		(void)fprintf(stderr, "\rusage: whoami\n\r");
	else
		(void)fprintf(stderr, "\r%s\n\r%s%s\n\r%s\n\r%s\n\r%s\n\r%s\n\r%s\n\r%s\n\r",
		    "usage: id [user]",
#ifdef USE_BSM_AUDIT
		    "       id -A\n\r",
#else
		    "",
#endif
		    "       id -F [user]",
		    "       id -G [-n] [user]",
		    "       id -M",
		    "       id -P [user]",
		    "       id -g [-nr] [user]",
		    "       id -p [user]",
		    "       id -u [-nr] [user]");
	// exit(1);
}
